src/tarball.rs

```diff
@@ -0,0 +1,166 @@
+//! Pin a tarball URL source
+//!
+//! Optionally (if the host supports it) can use the "Lockable HTTP Tarball Protocol" from flakes.
+//! Reference: <https://github.com/nixos/nix/blob/56763ff918eb308db23080e560ed2ea3e00c80a7/doc/manual/src/protocols/tarball-fetcher.md>
+
+use anyhow::{Context, Result};
+use reqwest::header::HeaderName;
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+use crate::*;
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+pub struct TarballPin {
+    /// URL provided as user input.
+    ///
+    /// Will get moved into `LockedTarball` once the Pin has been updated at
+    /// least once. This is a little bit of a hack but we need to be able to
+    /// update the `url` from within `TarballPin::update` so we have to store it
+    /// within `LockedTarball` which doesn't exist yet the first time `update`
+    /// is called.
+    #[serde(skip)]
+    pub url: Option<Url>,
```

Comment:
I very much do not like this, what are the alternative designs? Would it be possible to simply always store that url regardless of locking?

You have `TarballPin { url }` which always contains the input URL. Then you have `LockedTarball { Option<locked_url> }` if the `url` is lockable. The fetcher uses `locked_url` if present and `url` otherwise.

Reply:
yes, definitely agreed. i don't know why i got stuck in the mindset that we always have to put the url to download into `url`. adding a condition to the fetcher is much much simpler